### DISCUSSION TITLE
Feature request: anonymous messaging (reviewer <-> candidate)

### DISCUSSION THREAD

**Discussion by rommelfs** [pretalx/pretalx#1071] [Ideas]
I just started looking into this promising project and I like it already a lot.

And I have a feature request already:

CFP systems as far as I know are lacking the following feature. When a reviewer has a question regarding the submitted paper/content/proposal, he is required to 'know' or 'have' the contact details of the candidate (which can be impossible if reviewing anonymously). Then the reviewer has to send a mail in his name to the candidate, and reviews the answer to his mail address - breaking privacy and also might lead to missed mails or forgotten transfer into the CFP system.

It would be awesome if you could implement a kind of anonymous messaging system, where a reviewer can ask a question to the candidate. Question, answer and time to answer should be recorded within the review for all eligible reviewers to see. By default, the name of the reviewer shouldn't be revealed (unless otherwise expressed).

Bonus: a notification to the reviewer should be sent once an open question is answered.

Thanks for your consideration,
Sascha

**Comment by rixx** [comment]
While I appreciate the suggestion, and am going to leave it open to consider adding it later on, this has a problem: We'd need to handle incoming mails, and we'd basically be re-implementing ticket systems, which sounds like feature creep to me

**Comment by rommelfs** [comment]
Thanks for your quick feedback.

I can imagine the feature could be implemented in a quick-win fashion using the internals of the CFP system (meaning the user gets a notification about a new question by mail, has to log in, answers question in CFP system, reviewer is notified via mail about new answer). That would eliminate the (admitted) feature creep of a ticket request system re-implementation.

The anonymous part is (for me, personally) not the most important, but that's more a religious ^H^H^H^H^H^H^H^H^Hstrategic decision.

**Comment by rixx** [comment]
Doing this via in-system notification is definitely less feature-creepy, I'm just always a bit wary of re-implementing email or chat in my software. I can't promise a timeline for this, but it definitely looks like a feature worth including.

**Comment by rixx** [comment]
This is an issue that is well-suited for external contributors – let me know if I can be of any help, either via this issue or via support@pretalx.com.

## What this is about

Reviewers might have questions for speakers, but do not (and should not!) want to disclose their identity, or do not even know who the speaker is if the event runs anonymised reviews. They should have the possiblity of asking questions for clarification of the speaker, and the speaker should be able to respond.

## How to get started

First of all, leave a comment on this issue to make sure nobody else starts working on this issue at the same time!

Next, you need pretalx running locally on your machine. First, [fork](https://github.com/pretalx/pretalx/fork) the repository to your own GitHub account. Then clone it locally (or [set it as your remote](https://help.github.com/articles/configuring-a-remote-for-a-fork/) if you already cloned it earlier)

Follow the [developer setup guide](https://docs.pretalx.org/en/latest/developer/setup.html), and then play around with pretalx a bit.
Have a look at the places impacting this issue. Look at the code responsible for the issues (searching for relevant words or files should help you find your way around). We have an explanation of our [code structure](https://docs.pretalx.org/en/latest/developer/concepts.html) and the actual [data models](https://docs.pretalx.org/en/latest/developer/interfaces/models.html).


## How to do this

We will need a new model for this. It would go in `pretalx/submission/models/review.py`, and would look roughly like this:

```python
class ReviewMessage(…):
    submission = ForeignKey('submission.Submission')
    author = ForeignKey('person.User')
    created = DateTimeField(auto_now_add=True)
    direction = CharField(choices=(('from_review', _('From reviewers')), ('to_review', _('To reviewers'))))
    subject = CharField()
    text = CharField()
    reply_to = ForeignKey('self')
```

Reviewers should be able to see each others' messages to avoid duplicate requests. The reply_to key might not be necessary, we can just display messages in their order, too.

You would then need to add a form and logic to the review interface to allow reviewers to send messages and display all messages sent and received. When a message is sent, all speakers for that submission should receive an email containing the question and a link to the page where they can be answered (ideally, the submission page).

## Submit the PR

There are a number of tutorials on the Pull Request workflow online, for example this [short version](https://guides.github.com/introduction/flow/) by GitHub itself and this [more in-depth tutorial](https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github).

Remember to provide documentation and tests, if applicable. Run the tests before opening your PR, including style checks. We have a [contribution guide](https://docs.pretalx.org/en/latest/developer/contributing.html), which ought to prove helpful at this stage.
